programs/clearing_house/src/lib.rs

```diff
@@ -754,6 +757,226 @@ pub mod clearing_house {
         Ok(())
     }
 
+    #[access_control(
+        exchange_not_paused(&ctx.accounts.state)
+    )]
+    pub fn settle_lp<'info>(ctx: Context<SettleLP>, _market_index: u64) -> Result<()> {
+        // not ready yet
+        panic!("Settle LP not ready yet...");
+
+        /*
+        let user = &mut load_mut(&ctx.accounts.user)?;
+        let remaining_accounts_iter = &mut ctx.remaining_accounts.iter().peekable();
+
+        let market_map = MarketMap::load(
+            &get_writable_markets(market_index),
+            &MarketOracles::new(),
+            remaining_accounts_iter,
+        )?;
+        let mut market = market_map.get_ref_mut(&market_index)?;
+
+        let position_index = get_position_index_lp(&user.positions, market_index)?;
+        let lp_position = &mut user.positions[position_index];
+
+        settle_lp_position(lp_position, lp_position.lp_tokens, &mut market)?;
+
+        Ok(()) */
+    }
+
+    #[access_control(
+        exchange_not_paused(&ctx.accounts.state)
+    )]
+    pub fn remove_liquidity<'info>(
+        ctx: Context<AddRemoveLiquidity>,
+        market_index: u64,
+    ) -> Result<()> {
+        let user = &mut load_mut(&ctx.accounts.user)?;
+        let remaining_accounts_iter = &mut ctx.remaining_accounts.iter().peekable();
+
+        let market_map = MarketMap::load(
+            &get_writable_markets(market_index),
+            &MarketOracles::new(),
+            remaining_accounts_iter,
+        )?;
+        let mut market = market_map.get_ref_mut(&market_index)?;
+
+        // need position_lp bc could get incorrect account otherwise (old account with unsettled_pnl)
+        let position_index = get_position_index_lp(&user.positions, market_index)?;
+        let lp_position = &mut user.positions[position_index];
+
+        let lp_tokens_to_burn = lp_position.lp_tokens; // tmp
+
+        if lp_tokens_to_burn == 0 {
+            return Ok(());
+        }
+
+        validate!(
+            lp_position.lp_tokens >= lp_tokens_to_burn,
+            ErrorCode::InsufficientLPTokens,
+            "Trying to burn more lp tokens than the user has",
+        )?;
+
+        // settle the lp first
+        settle_lp_position(lp_position, lp_tokens_to_burn, &mut market)?;
+
+        // transform lp_position into a market position
+        lp_position.lp_tokens = lp_position
+            .lp_tokens
+            .checked_sub(lp_tokens_to_burn)
+            .ok_or_else(math_error!())?; // burn tokens
+
+        // update funding rate to the markets position
+        lp_position.last_cumulative_funding_rate = if lp_position.base_asset_amount > 0 {
+            market.amm.cumulative_funding_rate_long
+        } else {
+            market.amm.cumulative_funding_rate_short
+        };
+
+        // track new position in the market
+        // TODO: probably want to refactor all this into one fcn (auction PR fcn?)
```

Comment:
check out `update_position_and_market` on master, handles update user position and market